Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -329,56 +328,92 @@ TextureCache::TCacheEntryBase* TextureCache::Load(const u32 stage)
+	std::pair <TexCache::iterator, TexCache::iterator> iter_range = textures.equal_range(address);
+	TexCache::iterator iter = iter_range.first;
+	if (iter != iter_range.second)
 	{
-		// 1. Calculate reference hash:
-		// calculated from RAM texture data for normal textures. Hashes for paletted textures are modified by tlut_hash. 0 for virtual EFB copies.
-		if (g_ActiveConfig.bCopyEFBToTexture && entry->IsEfbCopy())
-			tex_hash = TEXHASH_INVALID;
-
-		// 2. a) For EFB copies, only the hash and the texture address need to match
-		if (entry->IsEfbCopy() && tex_hash == entry->hash && address == entry->addr)
+		if (iter->second->IsEfbCopy())
 		{
-			// TODO: Print a warning if the format changes! In this case,
-			// we could reinterpret the internal texture object data to the new pixel format
-			// (similar to what is already being done in Renderer::ReinterpretPixelFormat())
-			return ReturnEntry(stage, entry);
+			// For EFB copies, only the hash and the texture address need to match. Ignore the hash when
+			// using EFB to texture, because there's no hash in this case
+			if (g_ActiveConfig.bCopyEFBToTexture || tex_hash == iter->second->hash)
+			{
+				// TODO: Print a warning if the format changes! In this case,
+				// we could reinterpret the internal texture object data to the new pixel format
+				// (similar to what is already being done in Renderer::ReinterpretPixelFormat())
+				// TODO: Convert paletted textures, which are efb copies, using the right palette, so they display correctly
+				return ReturnEntry(stage, iter->second);
+			}
+			else
+			{
+				// pool this texture and make a new one later
+				FreeTexture(iter->second);
+				textures.erase(iter);
+			}
 		}
-
-		// 2. b) For normal textures, all texture parameters need to match
-		if (address == entry->addr && tex_hash == entry->hash && full_format == entry->format &&
-			entry->native_levels >= tex_levels && entry->native_width == nativeW && entry->native_height == nativeH)
+		else
 		{
-			return ReturnEntry(stage, entry);
-		}
+			TexCache::iterator oldest_entry = iter;
+			int temp_frameCount = 0x7fffffff;
+
+			while (iter != iter_range.second)
+			{
+				// For normal textures, all texture parameters need to match
+				if (tex_hash == iter->second->hash && full_format == iter->second->format && iter->second->native_levels >= tex_levels &&
+					iter->second->native_width == nativeW && iter->second->native_height == nativeH)
+				{
+					return ReturnEntry(stage, iter->second);
+				}
+
+				// Find the entry which hasn't been used for the longest time
+				if (iter->second->frameCount != FRAMECOUNT_INVALID && iter->second->frameCount < temp_frameCount)
+				{
+					temp_frameCount = iter->second->frameCount;
+					oldest_entry = iter;
+				}
+				++iter;
+			}
 
-		// pool this texture and make a new one later
-		FreeTexture(entry);
+			// If at least one entry was not used for the same frame, overwrite the oldest one
+			if (temp_frameCount != 0x7fffffff)
```

Comment:
All of this framecount handling is a bad idea, but I have no better one. But imo this is out of scope for this PR ;)